[ENH] Generate IDs when not given in  add

spikechroma opened the pull request:
## Description of changes

*Summarize the changes made by this PR.*
 - Improvements & Bug fixes
	 - ...
 - New functionality
	 - when a user uses add on an collection, they are no longer required to pass in an array of IDs. They will be automatically generated if not given on the server side. 

## Test plan
*How are these changes tested?*

- [ ] Tests pass locally with `pytest` for python, `yarn test` for js, `cargo test` for rust

## Documentation Changes
*Are all docstrings for user-facing APIs updated if required? Do we need to make documentation changes in the [docs repository](https://github.com/chroma-core/docs)?*

spikechroma commented:
* **#2699** <a href="https://app.graphite.dev/github/pr/chroma-core/chroma/2699?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `spike/generate_ids`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikechroma and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

atroyn commented on chromadb/api/types.py:
```diff
@@ -292,6 +296,10 @@ def validate_ids(ids: IDs) -> IDs:
         if not isinstance(id_, str):
             raise ValueError(f"Expected ID to be a str, got {id_}")
+
+        if len(id_) == 0:
```
On the fence about this still.

spikechroma replied:
we can discuss offline

atroyn replied:
The empty string is a valid identifier, and would still be unique - I have no idea why anyone would want to do this, but let's let them.

atroyn commented on chromadb/api/models/CollectionCommon.py:
```diff
@@ -181,7 +181,7 @@ def _unpack_embedding_set(
     def _validate_embedding_set(
         self,
-        ids: IDs,
+        ids: Optional[IDs],
```
I don't think `ids` can be `Optional` here, since `validate_ids` fails on `None`

spikechroma replied:
true. however, if we are generating ids on the server level, ids should be allowed to be optional here

atroyn commented on chromadb/api/models/Collection.py:
```diff
@@ -82,14 +83,18 @@ def add(
 
         self._client._add(
-            embedding_set["ids"],
+            cast(IDs, embedding_set["ids"]),
```
Why this `cast` ? `ids` on `EmbeddingSet` aren't `Optional`

spikechroma replied:
true. updated

atroyn commented on chromadb/api/models/AsyncCollection.py:
```diff
@@ -75,14 +76,18 @@ async def add(
 
         await self._client._add(
-            embedding_set["ids"],
+            cast(IDs, embedding_set["ids"]),
```
Why this `cast` ?

atroyn commented on chromadb/api/models/AsyncCollection.py:
```diff
@@ -297,13 +302,13 @@ async def upsert(
         await self._client._upsert(
             collection_id=self.id,
-            ids=embedding_set["ids"],
+            ids=cast(IDs, embedding_set["ids"]),
```
why this `cast` ?

spikechroma replied:
removed

atroyn reviewed (commented):
We discussed offline about moving the ID generation down into the server logic, and whether that would mean having to move the validators further down. 

I think the right thing is to move both down into the server; we may want to make the validation logic change before making this one. Let's discuss further what the right answer might be.

atroyn reviewed (commented):
We discussed offline about moving the ID generation down into the server logic, and whether that would mean having to move the validators further down. 

I think the right thing is to move both down into the server; we may want to make the validation logic change before making this one. Let's discuss further what the right answer might be.

HammadB commented on chromadb/test/property/strategies.py:
```diff
@@ -458,17 +458,26 @@ def recordsets(
     )
 
+    # This probablistic event is used to mimic user behavior when they don't provide ids
+    if for_add and np.random.rand() < 0.5:
```
its preferred not to use np.random.rand() in strategies, please use draw(st.bools)

HammadB commented on chromadb/server/fastapi/__init__.py:
```diff
@@ -744,6 +745,34 @@ async def delete_collection(
+
+        generated_ids = []
+        for _ in range(n):
+            generated_ids.append(str(uuid4()))
```
everytime a uuid is turned into a string a small piece of me dies

HammadB commented on chromadb/test/property/strategies.py:
```diff
@@ -458,17 +458,26 @@ def recordsets(
     min_metadata_size: int = 0,
     max_metadata_size: Optional[int] = None,
+    # ids can only be optional for add operations
+    for_add: bool = False,
```
i think this parameter name is confusing and underspecified - can we clarify it or somehow make this less option-heavy?

HammadB commented on chromadb/test/test_api.py:
```diff
@@ -1207,6 +1207,11 @@ def test_invalid_id(client):
         collection.add(embeddings=[0, 0, 0], ids=[1], metadatas=[{}])
     assert "ID" in str(e.value)
 
+    # Upsert with an empty id
```
test api is a kitchen sink we should leave untouched. Please move the test out of this file.

I like the basic unit tests but would prefer to see basic tests that cover cases like

- No ids supplied - and the generated ids when retrieved with get() return what you expect
- Partial ids supplied fails
- If you supply ids, the expected result is returned

spikechroma replied:
added

HammadB commented on chromadb/server/fastapi/__init__.py:
```diff
@@ -744,6 +745,34 @@ async def delete_collection(
+    @staticmethod
+    def _generate_ids_when_not_present(
+        ids: Optional[List[str]],
+        documents: Optional[List[Optional[str]]],
```
nit: why do you need to pass docs/urs/embeddings into this, it makes more sense to me that this would take the value N that it needs

HammadB commented on chromadb/server/fastapi/__init__.py:
```diff
@@ -744,6 +745,34 @@ async def delete_collection(
         )
 
+    @staticmethod
+    def _generate_ids_when_not_present(
```
I don't think this method can live here? What happens in local chroma?

HammadB replied:
<img width="1295" alt="Screenshot 2024-08-22 at 12 48 33 AM" src="https://github.com/user-attachments/assets/5a00dc91-53d2-499a-a861-571dc2cebc44">

HammadB replied:
Yeah I don't think this change is fully considered, not sure if its halfway but this has to live on layer lower if you want it at the "conceptual server" layer.

Would also like to have testing that would catch this as opposed to me reading it.

HammadB reviewed (changes requested):
I am not sure if this change is partways, but as is this change can't be merged as its broken on single node. Putting the id generation in server/fastapi is not the right way to get this to be uniform across local/deployed single node. We should put the change at the level of "ServerAPI" so that it runs at the conceptual server layer. I am happy to go over the code architecture tomorrow to make this clearer / discuss more. 

I also think we need more testing against this functionality since its additive and doesn't break existing tests. The property test modifications are good, and I think basic point unit tests are warranted too.

HammadB commented:
This addresses https://github.com/chroma-core/chroma/issues/164 in part, where we agreed validation duplication is preferred. 

cc @levand @jeffchuber

spikechroma commented:
> This stack addresses #164 in part, where we agreed validation duplication is preferred.
> 
> cc @levand @jeffchuber

I addressed all the comments

atroyn commented on chromadb/api/types.py:
```diff
@@ -69,7 +69,7 @@ def maybe_cast_one_to_many_embedding(
         return None
 
-    if isinstance(target, List):
+    if isinstance(target, List) and len(target) != 0:
```
We should not have `Embeddings` of length 0

atroyn commented on chromadb/api/fastapi.py:
```diff
@@ -413,15 +414,29 @@ def _add(
+            "post",
+            "/collections/" + str(collection_id) + "/add",
+            json={
+                "ids": batch[0],
```
Since this gets passed to the segment, shouldn't it be possible for `ids` to be `None` here?

atroyn commented on chromadb/api/async_fastapi.py:
```diff
@@ -448,11 +449,25 @@ async def _add(
+            "post",
+            "/collections/" + str(collection_id) + "/add",
+            json={
+                "ids": batch[0],
```
`ids` can be `None` here since we handle this in the Segment?

atroyn commented on chromadb/api/async_client.py:
```diff
@@ -266,7 +267,7 @@ async def _add(
         documents: Optional[Documents] = None,
         uris: Optional[URIs] = None,
-    ) -> bool:
+    ) -> AddResult:
```
`ids` can be `None` here?

atroyn commented on chromadb/api/models/CollectionCommon.py:
```diff
@@ -386,7 +386,7 @@ def _process_add_request(
     ) -> RecordSet:
         unpacked_embedding_set = self._unpack_embedding_set(
-            ids,
+            ids if ids is not None else [],
```
We shouldn't use the empty list as a placeholder

atroyn commented on chromadb/api/segment.py:
```diff
@@ -814,6 +826,30 @@ def get_settings(self) -> Settings:
+        n_uris: int,
+        n_embeddings: int,
+    ) -> IDs:
+        if ids is not None and len(ids) != 0:
```
We should never pass a list of length 0

atroyn commented on chromadb/api/segment.py:
```diff
@@ -339,11 +344,18 @@ def _add(
 
+        ids = self.generate_ids_when_not_present(
+            ids=ids,
+            n_documents=len(documents) if documents is not None else 0,
```
should just compute this length inside the generator function itself - this encapsulates handling of `None` and allows us to be defensive.

atroyn commented on chromadb/test/property/strategies.py:
```diff
@@ -461,14 +467,21 @@ def recordsets(
+    n_records = len(ids)
+
+    if len(ids) == 0:
+        n_records += 1
```
we will only ever create one record if we draw 0 ids? I think it's better to draw `n` first, then use that length for everything else.

HammadB replied:
Yeah agreed, also in this case you end up only generating ids if there are 0 ids right, i think we should draw n and then flip on ids

atroyn commented on chromadb/test/property/strategies.py:
```diff
@@ -553,7 +568,7 @@ def where_clause(draw: st.DrawFn, collection: Collection) -> types.Where:
         legal_ops: List[Optional[str]] = [None, "$eq"]
     else:
-        legal_ops: List[Optional[str]] = [None, "$eq", "$ne", "$in", "$nin"]
+        legal_ops = [None, "$eq", "$ne", "$in", "$nin"]
```
What happened to this type annotation?

atroyn reviewed (commented):
There's a few things I think we need to change before we can land this:

- We should be able to pass `None` in the `ids` field right down to the segment level; right now we can pass `None` at the client and at the segment, but not in between. I think that's why we're passing around an empty list for IDs sometimes as a placeholder, which we shouldn't do. 

- Test implementation currently generates only 1 record if we don't generate any IDs. We should instead choose to generate `n` records, then decide whether we generate any IDs. Otherwise, we only ever exercise the path where we don't get a list of IDs with one record.

atroyn reviewed (changes requested):
There's a few things I think we need to change before we can land this:

- We should be able to pass `None` in the `ids` field right down to the segment level; right now we can pass `None` at the client and at the segment, but not in between. I think that's why we're passing around an empty list for IDs sometimes as a placeholder, which we shouldn't do. 

- Test implementation currently generates only 1 record if we don't generate any IDs. We should instead choose to generate `n` records, then decide whether we generate any IDs. Otherwise, we only ever exercise the path where we don't get a list of IDs with one record.

atroyn commented:
Now that we generate IDs, we should also check that we generate the right amount. Additionally, I think we should check that ids returned from `add` are returned in the right order - these should be additional asserts in the property tests.

HammadB commented on chromadb/test/property/strategies.py:
```diff
@@ -408,6 +412,8 @@ def document(draw: st.DrawFn, collection: Collection) -> types.Document:
     # We also don't want them to contain certan special
     # characters like _ and % that implicitly involve searching for a regex in sqlite.
+
+    blacklist_categories: Tuple[str, ...] = ()
```
nit: why define this as an empty tuple and redefine it below, we can just define it correctly the first time

spikechroma replied:
fixed

HammadB commented on chromadb/test/property/test_add.py:
```diff
@@ -27,7 +27,7 @@
 @given(
     collection=collection_st,
-    record_set=strategies.recordsets(collection_st, min_size=1, max_size=500),
+    record_set=strategies.recordsets(collection_st, min_size=0, max_size=500),
```
i find this change smelly - we are overloading the meaning of 0 to sometimes generate ids, but the caller of this function her never wants 0

HammadB commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -293,7 +293,7 @@ def persist_generated_data_with_old_version(
 @given(
     collection_strategy=collection_st,
-    embeddings_strategy=strategies.recordsets(collection_st),
+    embeddings_strategy=strategies.recordsets(collection_st, min_size=1),
```
Also min_size was defaulting to 1 so does this really change anything?

HammadB commented on clients/js/src/ChromaClient.ts:
```diff
@@ -2,7 +2,6 @@ import { AdminClient } from "./AdminClient";
 import { authOptionsToAuthProvider, ClientAuthProvider } from "./auth";
```
Please have @codetheweb review JS changes

codetheweb replied:
(@spikechroma lmk when https://github.com/chroma-core/chroma/pull/2714 lands and you rebase on that, I'll review then)

spikechroma replied:
sg

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -190,11 +192,9 @@ def test_add_large(
-        and should_compact
-        and len(normalized_record_set["ids"]) > 10
-    ):
+    n_ids = len(normalized_record_set["ids"] or [])
```
this breaks the count invariant

atroyn replied:
we call `add` on the collectiom above, which means we will always have `ids`

atroyn commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -267,7 +267,9 @@ def persist_generated_data_with_old_version(
         actual_ids = result["ids"]
-        embedding_id_to_index = {id: i for i, id in enumerate(check_embeddings["ids"])}
+        embedding_id_to_index = {
+            id: i for i, id in enumerate(check_embeddings["ids"] or [])
```
This breaks the test

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -138,18 +138,24 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
         self.on_state_change(EmbeddingStateMachineStates.delete_by_ids)
-        indices_to_remove = [self.record_set_state["ids"].index(id) for id in ids]
+
+        normalized_ids = self.record_set_state["ids"] or []
```
By the time we arrive at `delete` we should know the ids of what we intend to delete

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -114,18 +114,18 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
         )
 
-        if len(normalized_record_set["ids"]) > 0:
+        ids = normalized_record_set["ids"] or []
```
This logic would be clearer if we were clear about the possible branches:
- There are already IDs in the `record_set_state` and the new `record_set` IDs are not `None` - possible intersection of IDs
- There are no IDs in the `record_set_state` and the `record_set` IDs are not `None` - purely additive, with IDs present
- There are already IDs in the `record_set_state` and the `record_set` IDs are `None` - purely additive, IDs are generate during `add`
- There are no IDs in the `record_set_state` and the `record_set` IDs are `None` - purely additive, IDs are generated during `add`

We should restructure this to make it clear.

spikechroma replied:
resolved offline

spikechroma replied:
i will leave it as it is since found intersected ids vs not found intersected ids is easier to understand than the 4 scenarios imo

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -138,18 +138,24 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
             self.collection.add(**normalized_record_set)  # type: ignore[arg-type]
             self._upsert_embeddings(cast(strategies.RecordSet, filtered_record_set))
-            return multiple(*filtered_record_set["ids"])
+            return multiple(*filtered_record_set["ids"] or [])
```
`filtered_record_set` shouuld naturally return the right thing here without need for `[]`

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -138,18 +138,24 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
+
             self._upsert_embeddings(cast(strategies.RecordSet, normalized_record_set))
-            return multiple(*normalized_record_set["ids"])
+            return multiple(*normalized_record_set["ids"] or [])
```
?

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -173,7 +179,7 @@ def update_embeddings(self, record_set: strategies.RecordSet) -> None:
 
     # Using a value < 3 causes more retries and lowers the number of valid samples
-    @precondition(lambda self: len(self.record_set_state["ids"]) >= 3)
+    @precondition(lambda self: len(self.record_set_state["ids"] or []) >= 3)
```
?

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -229,9 +235,16 @@ def _upsert_embeddings(self, record_set: strategies.RecordSet) -> None:
             # Update path
-            if id in self.record_set_state["ids"]:
+            if (
+                self.record_set_state["ids"] is not None
```
internal state should never be `None`

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -366,21 +379,24 @@ def wait_for_compaction(self) -> None:
             record_set
         )
+
+        ids = normalized_record_set["ids"] or []
```
By now we have the ids from `add_embeddings` even if the ids on `record_set` are generated `None` - we have them in `res`

atroyn reviewed (changes requested):
Some cleanup needed in internal logic, and tests need to be fixed up.

The test logic is currently wrong in several places, and tests are getting overriden / short circuited. We should be careful to correctly reason about the case where we generate a record_set which can have `None` for its IDs. We can also avoid generating `None` ids for state transitions which don't accept them, e.g. `upsert` / `update`.

spikechroma commented on chromadb/api/types.py:
```diff
@@ -62,7 +62,7 @@ def maybe_cast_one_to_many_embedding(
     if isinstance(target, List):
         # One Embedding
-        if isinstance(target[0], (int, float)):
+        if len(target) > 0 and isinstance(target[0], (int, float)):
```
leave it as it is according to discussions

atroyn replied:
I think we should just be defensive and raise here.

spikechroma commented on chromadb/api/models/CollectionCommon.py:
```diff
@@ -184,12 +185,12 @@ def _validate_record_set(
-            and record_set["documents"] is None
-            and record_set["images"] is None
-            and record_set["uris"] is None
+            and (record_set["embeddings"] is None or len(record_set["embeddings"]) == 0)
```
leave it as it is according to discussions

spikechroma commented on chromadb/test/property/strategies.py:
```diff
@@ -194,7 +194,7 @@ def create_embeddings_ndarray(
     count: int,
     dtype: npt.DTypeLike,
-) -> np.typing.NDArray[Any]:
+) -> npt.NDArray[Any]:
```
https://github.com/chroma-core/chroma/issues/2737

spikechroma commented on chromadb/test/property/test_add.py:
```diff
@@ -99,37 +101,37 @@ def _test_add(
+        normalized_record_set["ids"] = result["ids"]
+        n_ids = len(result["ids"])
+
+    if not NOT_CLUSTER_ONLY and should_compact and n_ids > 10:
```
https://github.com/chroma-core/chroma/issues/2738

spikechroma commented on chromadb/test/property/strategies.py:
```diff
@@ -458,17 +462,25 @@ def recordsets(
     collection = draw(collection_strategy)
 
-    ids = list(
+    ids: Optional[List[types.ID]] = list(
```
resolved offline

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -99,37 +101,41 @@ def _test_add(
-    coll.add(**record_set)  # type: ignore
+    result = coll.add(**record_set)  # type: ignore
+
+    n_ids = (
```
This is weird - we don't actually need `n_ids` - we are just checking for the total number of records. Let's use the function we wrote for getting that length here instead.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -179,22 +185,29 @@ def test_add_large(
     ):
-        coll.add(*batch)
+        result = coll.add(*batch)
+        if result["ids"] is not None:
```
result["ids"] should never be `None`; we can set the type on result as `result: IDs` or else cast, or else raise.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -179,22 +185,29 @@ def test_add_large(
+        if result["ids"] is not None:
+            ids.extend(result["ids"])
+
+    n_ids = (
```
we are again looking for the number of records, not `n_ids` specifically. Let's re-use our counting logic.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -179,22 +185,29 @@ def test_add_large(
-    initial_version = coll.get_model()["version"]
+    initial_version = cast(int, coll.get_model()["version"])
 
+    ids = []
```
in `create_large_recordset`, `ids` is never `None` - if it were, `create_batches` would fail because it would call `len(None)`. Let's just be defensive and raise if `ids` turns out to be `None` on `RecordSet` in this case.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -179,22 +185,29 @@ def test_add_large(
-        and should_compact
-        and len(normalized_record_set["ids"]) > 10
-    ):
+    if not NOT_CLUSTER_ONLY and should_compact and n_ids > 10:
```
Let's create an issue to fold the `not NOT_CLUSTER_ONLY and should_compact` logic into `wait_for_version_increase` everywhere.

spikechroma replied:
https://github.com/chroma-core/chroma/issues/2738 did this yesterday

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -287,3 +300,71 @@ def test_add_partial(client: ClientAPI) -> None:
+
+
+def test_add_with_no_ids(client: ClientAPI) -> None:
+    """Tests adding a record set with some of the fields set to None."""
```
'some of the fields' ?

atroyn replied:
I don't think these belong in property tests - we should remove these and put them elsewhere.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -287,3 +300,71 @@ def test_add_partial(client: ClientAPI) -> None:
+    assert len(results["ids"]) == 6
+
+
+def test_add_with_partial_ids(client: ClientAPI) -> None:
```
Let's move these out of here as part of the `test_api` refactor

atroyn commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -256,18 +256,24 @@ def persist_generated_data_with_old_version(
             embedding_function=not_implemented_ef(),
         )
-        coll.add(**embeddings_strategy)
+        result = coll.add(**embeddings_strategy)
```
- Old versions will not return a `result` 
- Old versions can't accept a `RecordSet` with null IDs. 

We shold determine which version we are in when `add` is called and do the right thing accordingly.

atroyn replied:
I'm not sure how this test is currently passing.

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -136,20 +140,31 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
-        indices_to_remove = [self.record_set_state["ids"].index(id) for id in ids]
+
+        normalized_ids = self.record_set_state["ids"]
+        if normalized_ids is None:
```
Should raise instead.

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -136,20 +140,31 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
         self.on_state_change(EmbeddingStateMachineStates.delete_by_ids)
-        indices_to_remove = [self.record_set_state["ids"].index(id) for id in ids]
+
+        normalized_ids = self.record_set_state["ids"]
```
what makes this 'normalized' ?

spikechroma replied:
changed to state_ids

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -229,6 +248,10 @@ def _upsert_embeddings(self, record_set: strategies.RecordSet) -> None:
             record_set
         )
+
+        if normalized_record_set["ids"] is None:
```
should raise

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -366,21 +389,29 @@ def wait_for_compaction(self) -> None:
         )
+
+        ids = normalized_record_set["ids"]
+        n_ids = len(invariants.wrap(res))
```
confused, let's discuss

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -407,16 +438,29 @@ def delete_by_ids(self, ids: IDs) -> None:
     def update_embeddings(self, record_set: strategies.RecordSet) -> None:
         super().update_embeddings(record_set)
+
+        n = (
```
`ids` should never be `None` here; should raise

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -427,14 +471,20 @@ def update_embeddings(self, record_set: strategies.RecordSet) -> None:
     def upsert_embeddings(self, record_set: strategies.RecordSet) -> None:
         super().upsert_embeddings(record_set)
+
+        normalized_ids = (
```
the strategies we use to generate record sets for `update` and `upsert` have `can_ids_be_none` set to `False` - we should raise if we ever do.

atroyn commented on chromadb/api/types.py:
```diff
@@ -515,39 +524,46 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
         )
 
 
+def get_n_items_from_record_set(record_set: RecordSet) -> Tuple[str, int]:
```
we should re-use this as much as we can!

atroyn commented on chromadb/api/types.py:
```diff
@@ -515,39 +524,46 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
-            continue
+    n_items_field, n_items = get_n_items_from_record_set(record_set)
+    if n_items == 0:
+        return
```
how did we get an empty `RecordSet` ? - based on how `get_n_items_from_record_set` works, this must be an empty list - We should probably raise.

atroyn commented on chromadb/api/types.py:
```diff
@@ -515,39 +524,46 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
+            if n != n_items:
                 raise ValueError(
-                    f"Number of {field} ({n}) does not match number of ids ({n_ids})"
+                    f"Number of {field} ({n}) does not match number of {n_items_field} ({n_items})"
```
we should go over everything before we raise, just in case they're all different lengths, so that the error message can be maximally informative.

atroyn commented on chromadb/api/segment.py:
```diff
@@ -336,27 +338,32 @@ def delete_collection(
+        ids = self.generate_ids_when_not_present(record_set)
+        record_set["ids"] = ids
+
         self._validate_recrod_set(
```
typo, `recrod`

atroyn commented on chromadb/api/segment.py:
```diff
@@ -836,6 +843,22 @@ def get_settings(self) -> Settings:
+
+        (_, n) = get_n_items_from_record_set(record_set)
+
+        generated_ids: List[str] = []
```
```suggestion
        generated_ids: List[str] = [str(uuid4()) for _ in range(n)]
```

atroyn reviewed (changes requested):
We should also be defensive and raise if we get `None` passed in as `ids` for `update` or `upsert`, at both client and server level.

atroyn commented on chromadb/test/property/test_add.py:
```diff
@@ -188,13 +189,12 @@ def test_add_large(
     ):
-        coll.add(*batch)
+        results = coll.add(*batch)
+        if results["ids"] is None:
```
Can `results["ids"]` ever be `None` ? I'm not sure this check is doing the right thing.

spikechroma replied:
removed

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -142,20 +146,31 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
-        indices_to_remove = [self.record_set_state["ids"].index(id) for id in ids]
+
+        state_ids = self.record_set_state["ids"]
+        if state_ids is None:
```
We should error before assignment.

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -372,23 +391,30 @@ def wait_for_compaction(self) -> None:
         normalized_record_set: strategies.NormalizedRecordSet = invariants.wrap_all(
             record_set
         )
+
```
Looking at this logic again, I'm confused about this change. Previously everything was from the length of the record_set ...

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -372,23 +391,30 @@ def wait_for_compaction(self) -> None:
+
+        self.log_operation_count += n_records
+
+        for id in res:
```
Need to make sure this is correct. @sanketkedia - not sure that this logic is doing what we expect.

atroyn reviewed (changes requested):
Some more consolidation and clarity, also need to check the correctness in a couple of places. 

Let's also move the javascript changes to another branch on this stack above this one, so we can look at it separately.

spikechroma commented on chromadb/api/types.py:
```diff
@@ -62,13 +62,12 @@ def maybe_cast_one_to_many_embedding(
-    # Already a sequence
 
+    # Already a sequence
     if isinstance(target, np.ndarray):
```
Need more clarification on this one. What do you mean that this won't do the right thing?

atroyn replied:
What happens if `target` is a numpy array of size (1,n) ?

atroyn commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -256,18 +256,22 @@ def persist_generated_data_with_old_version(
-        coll.add(**embeddings_strategy)
+        result = coll.add(**embeddings_strategy)
+
+        if result is not None and "ids" in result:
```
We should check if `embeddings_strategy["ids"]` is `None` before we do this assignment.

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -120,18 +120,22 @@ def add_embeddings(self, record_set: strategies.RecordSet) -> MultipleResults[ID
-            new_ids = list(set(normalized_record_set["ids"]).difference(intersection))
-            indices = [normalized_record_set["ids"].index(id) for id in new_ids]
+            new_ids = list(set(ids).difference(intersection))  # type: ignore[arg-type]
+            indices = [ids.index(id) for id in new_ids]  # type: ignore
```
No generic `#type: ignore`

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -197,8 +212,14 @@ def upsert_embeddings(self, record_set: strategies.RecordSet) -> None:
         invariants.count(
-            self.collection, cast(strategies.RecordSet, self.record_set_state)
+            self.collection,
+            cast(strategies.RecordSet, self.record_set_state),
```
Shouldn't `count` just take the new `RecordSetState` type instead?

spikechroma replied:
make a note

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -413,13 +454,18 @@ def delete_by_ids(self, ids: IDs) -> None:
+
+        if record_set["ids"] is None:
+            raise ValueError("IDs should not be empty")
+        n = len(invariants.wrap(record_set["ids"]))
```
Should we use the record set count method?

spikechroma replied:
we rely on the number of ids to get the number of items within update and upsert

atroyn commented on chromadb/test/property/test_embeddings.py:
```diff
@@ -433,14 +479,21 @@ def update_embeddings(self, record_set: strategies.RecordSet) -> None:
+        if record_set["ids"] is None:
+            raise ValueError("IDs should not be empty")
+
+        ids = invariants.wrap(record_set["ids"])
```
Should we use the record set count method?

spikechroma replied:
we use rely on the number of ids to get the number of items within update and upsert

atroyn commented on chromadb/api/types.py:
```diff
@@ -533,39 +544,91 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
+    error_messages = []
+    field_record_counts = []
+    count = 0
+    consistentcy_error = False
```
spelling?

spikechroma commented on chromadb/api/types.py:
```diff
@@ -519,39 +528,91 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
+    consistentcy_error = False
+
+    for field, value in record_set.items():
+        if value is None:
```
updated to check for all None scenario

spikechroma commented on chromadb/api/types.py:
```diff
@@ -519,39 +524,53 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
+    Get the number of items in the record set.
+    """
+    for field, value in record_set.items():
+        if isinstance(value, list) and len(value) > 0:
```
added logic to check for all inputs are none in `validate_record_set_consistency`

atroyn commented on chromadb/test/property/strategies.py:
```diff
@@ -682,6 +702,9 @@ def filters(
     else:
         ids = recordset["ids"]
 
+    if ids is None:
```
The type of `ids` is explicitly `Optional` - one of these should be changed.

atroyn commented on chromadb/test/property/test_filtering.py:
```diff
@@ -206,7 +207,7 @@ def test_filterable_metadata_get(
-    initial_version = coll.get_model()["version"]
+    initial_version = cast(int, coll.get_model()["version"])
 
     coll.add(**record_set)
```
Let's make it explicit that we cannot generate RecordSet without IDs here

spikechroma replied:
done

atroyn commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -256,18 +256,28 @@ def persist_generated_data_with_old_version(
-        coll.add(**embeddings_strategy)
+        result = coll.add(**embeddings_strategy)
+
+        if embeddings_strategy["ids"] is None:
```
We should gate this on the version, similar to what we do with create collection above.

atroyn replied:
Should rename to match `RecordSet` type.

spikechroma replied:
done

atroyn commented on chromadb/test/property/test_cross_version_persist.py:
```diff
@@ -293,18 +303,29 @@ def persist_generated_data_with_old_version(
     # the previous versions
     version, settings = version_settings
+
+    # TODO: This conditionn is subject to change as we decide on whether we want
```
typo

atroyn reviewed (commented):
Some cleanup to go and I think this is ready. Need input from @sanketkedia on the log state in tests.

spikechroma commented on chromadb/api/types.py:
```diff
@@ -546,39 +561,89 @@ def validate_embeddings(embeddings: Embeddings) -> Embeddings:
+
+def get_n_items_from_record_set(
+    record_set: RecordSet,
+) -> Tuple[str, int]:
```
updated func

spikechroma commented on chromadb/test/property/invariants.py:
```diff
@@ -67,11 +71,73 @@ def wrap_all(record_set: RecordSet) -> NormalizedRecordSet:
 
 
+def get_n_items_from_record_set_state(state_record_set: StateMachineRecordSet) -> int:
+    normalized_record_set = wrap_all(cast(RecordSet, state_record_set))
```
good call

spikechroma commented on chromadb/test/property/invariants.py:
```diff
@@ -67,11 +71,73 @@ def wrap_all(record_set: RecordSet) -> NormalizedRecordSet:
+
+    # we need to replace empty lists with None within the record set state to use get_n_items_from_record_set
+    # get_n_items_from_record_set would throw an error if it encounters an empty list
+    record_set_with_empty_lists_replaced: types.RecordSet = {
```
done
